### PR TITLE
p2pool: Restart monerod only when needed and with proper args

### LINKED ISSUE #3911: monerod second instance started by GUI
Linux x64
Ubuntu
Monero GUI 0.17.3.2 Mining monerod p2pool second daemon

The case is as follows:

The monerod is started a boot time. NO default directories used!

My lmdb is 100% synchronized. When I start the Monero GUI it connects to my local node. So far so good. 

The GUI connects to the already running monerod flawless and all is fine.

But when I start the mining by the GUI a second instance of the monero daemon seems to  initiated. 
I selected p2pool mining with main.
This second daemon is started with the wrong directory for the database. Default settings are used.
That brings it to the negative effect that the complete lmdb will be created (downloaded) again.

The problem I described above is repeatable.

My question is:

Why is the Monero GUI trying to establish a second daemon even it has connected successfully to daemon already running?

### LINKED ISSUE #3935: p2pool mining excessive ram usage never reclaims reboots required
So, I've noticed and seem to have confirmed, if I need to ever restart the MONERO GUI WALLET p2pool miner due to changing thread intensity that means restarting the whole miner setup which also seems to mean restarting the monerod daemon in the node section which ultimately combines to wasted RAM usage I can't even find in the task manager to kill off. So there seems to be a memory process problem for sure. I'm going to need to reboot my computer now lol 55 GB/64 GB ram, like, this isn't going to work.

tl;dr; it appears monerod or p2pool process in monero gui wallet never gets closed completely so end up having wasted memory usage on the computer and end up requiring a reboot since these processes don't show up/kill easily enough especially on something like windows

EDIT : PS: it's probably also clear to most users that trying to restart the MONERO GUI WALLET p2pool miner that almost always it never starts up properly again that you need to restart the daemon in the node settings tab again, that'd be the other ease of use thing, like upon start mining button : KILL ALL p2pool/monerod instances (FREE UP RAM) and start monerod afterwards

so yeah mining via the gui wallet seems good for 24/7 rigs
but kinda bad for main multi-tasking workstations considering you may start/stop multiple times a day 
and eventually the RAM seems to get out of hand

### PR DESCRIPTION
**Pull request by devhyper**
Resolves #3911 and may resolve #3935

### CODE REVIEW

**Comment by selsta** [comment]
One question. If someone has their monerod managed through something like systemd, does systemd even allow it that a different program restarts `monerod` with different parameters?

**Comment by devhyper** [comment]
> One question. If someone has their monerod managed through something like systemd, does systemd even allow it that a different program restarts `monerod` with different parameters?

@selsta The current behavior is that monerod restarts every time p2pool mining is started. This PR will just reduce the amount of restarts. I think in the case of systemd, monerod will be restarted, but not under systemd anymore, it will be its own separate process. I can open a separate PR to detect if monerod is under systemd.

**Comment by devhyper** [comment]
Should be ready to go now, haven't tested yet though.

**Comment by devhyper** [comment]
@selsta Just tested it, it works.

**Comment by devhyper** [comment]
@selsta

**Comment by selsta** [comment]
@devhyper will try to get it merged

**Comment by plowsof** [comment]
I can't compile this: (planning to confirm systemd monerod also)
```
/home//dev/monero-gui/src/daemon/DaemonManager.cpp: In member function ‘bool DaemonManager::checkUnderSystemd()’:
/home//dev/monero-gui/src/daemon/DaemonManager.cpp:358:45: warning: ‘static int QProcess::execute(const QString&)’ is deprecated: Use QProcess::execute(const QString &program, const QStringList &arguments) instead [-Wdeprecated-declarations]
  358 |         int underSystemd = QProcess::execute("/bin/sh -c \"ps -eo pid,cgroup | grep " + pid + " | grep -q .service$\"");
In file included from /usr/include/x86_64-linux-gnu/qt5/QtCore/QProcess:1,
                 from /home//dev/monero-gui/src/daemon/DaemonManager.h:37,
                 from /home//dev/monero-gui/src/daemon/DaemonManager.cpp:29:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qprocess.h:265:16: note: declared here
  265 |     static int execute(const QString &command);
/home//dev/monero-gui/src/daemon/DaemonManager.cpp: In member function ‘QString DaemonManager::getArgs()’:
/home//dev/monero-gui/src/daemon/DaemonManager.cpp:367:17: error: no matching function for call to ‘DaemonManager::running(NetworkType::Type)’
  367 |     if (!running(NetworkType::MAINNET)) {
/home//dev/monero-gui/src/daemon/DaemonManager.cpp:239:6: note: candidate: ‘bool DaemonManager::running(NetworkType::Type, const QString&) const’
  239 | bool DaemonManager::running(NetworkType::Type nettype, const QString &dataDir) const
/home//dev/monero-gui/src/daemon/DaemonManager.cpp:239:6: note:   candidate expects 2 arguments, 1 provided
make[3]: *** [src/CMakeFiles/monero-wallet-gui.dir/build.make:391: src/CMakeFiles/monero-wallet-gui.dir/daemon/DaemonManager.cpp.o] Error 1
make[3]: Leaving directory '/home//dev/monero-gui/build'
make[2]: *** [CMakeFiles/Makefile2:4361: src/CMakeFiles/monero-wallet-gui.dir/all] Error 2
make[2]: Leaving directory '/home//dev/monero-gui/build'
make[1]: *** [Makefile:136: all] Error 2
make[1]: Leaving directory '/home//dev/monero-gui/build'
make: *** [Makefile:32: default] Error 2
```
I will also test linux but for now, moving onto windows:

Win 10 using [docker-windows-static](https://github.com/monero-project/monero-gui/actions/runs/3223990972)

Test: monerod args `--zmq-pub=tcp://127.0.0.1:18083` `--disable-dns-checkpoints`    
Expected: monerd will not restart / mining begins    
Result: pass    
<details>
  <summary>Video</summary>

[no_restart.webm](https://user-images.githubusercontent.com/77655812/209424543-05d23864-e22c-4a18-a257-bc3082f5ece4.webm)


</details>

Test: monerod args `--prune-blockchain` `--zmq-pub=tcp://127.0.0.1:18083`     
Expected: monerod will restart because `--disable-dns-checkpoints` not present    
Result: monerod restarts but custom arg `--prune-blockchain` not present    
<details>
  <summary>Video</summary>

[prune_zmq_fine.webm](https://user-images.githubusercontent.com/77655812/209424489-29185cac-4449-4446-b7b7-2a11cc7718d0.webm)

</details>

Test: no args    
Expected: monerod restarts with --zmq-pub and --disable-dns-checkpoints set    
Result: endless loop / not working     
<details>
  <summary>Video</summary>

[no_args_loop.webm](https://user-images.githubusercontent.com/77655812/209423460-3868e045-7cdf-4ec4-9760-6292be0383c6.webm)

</details>

Test: monerod with only arg `--prune-blockchain`       
Expected: monerod to restart with prune/zmq/disable-dns      
Result: endless loops / not working        
<details>
  <summary>Video</summary>

[prune_only_loop.webm](https://user-images.githubusercontent.com/77655812/209423450-b5afac4a-3705-4201-a42f-8100c975838b.webm)

</details>

Test: monerod with only arg `--zmq-pub=tcp://127.0.0.1:18083`      
Expected: monerod to restart with zmq+disable-dns       
Result: pass       
<details>
  <summary>Video</summary>

[fine_with_tcp_only.webm](https://user-images.githubusercontent.com/77655812/209423429-fd3f9220-bc32-45d3-a769-e3bedbf54049.webm)

</details>

**Comment by devhyper** [comment]
@plowsof Deprecation and compile errors fixed, most of the cases should be fixed as well. I'll test some more when I can.

**Comment by plowsof** [comment]
Thanks! linux compiles now. some issues:
Linux: (self compiled)
daemon args in gui: `--prune-blockchain --zmq-pub=tcp://127.0.0.1:18083 --disable-dns-checkpoints`
error seen after attempting to start mining:
```
Failed to parse arguments: unrecognised option '--detach,--prune-blockchain,--zmq-pub=tcp://127.0.0.1:18083,--disable-dns-checkpoints,--check-updates,disabled,--non-interactive,--max-concurrency,4'
```
Win 10 [docker-static](https://github.com/monero-project/monero-gui/actions/runs/3771898859): 
i could not get any tests to pass - im seeing an infinite loop on all. when you have time to give a 2nd opinion that would be much appreciated (as im not 100% confident in my windows setup)

**Comment by devhyper** [comment]
@plowsof This should fix it, I'm currently not able to test right now, will be able to test in a few days.

**Comment by plowsof** [comment]
Thanks! Only seeing the `Failed to parse arguments: option '--detach' cannot be specified more than once` error now on linux. I can review also in a few days no problem.

**Comment by plowsof** [comment]
@devhyper i have made the adjustments so all test cases pass. I will test some time tomorrow to confirm, then make a pull request to your branch with the commit. We can get final touches done / merged asap after then:   https://github.com/monero-project/monero-gui/commit/368e49682e7aead5d55dc594df6dbfe7f95e3c61

**Comment by plowsof** [comment]
Win 10 using [docker-windows-static](https://github.com/plowsof/monero-gui/actions/runs/3791065734)
Ive still not actually tested systemd, i believe tomorrow, and ill make the PR to your branch then also, excuse the delay.

Test: monerod args `--zmq-pub=tcp://127.0.0.1:18083` `--disable-dns-checkpoints`    
Expected: monerod will not restart / mining begins    
Result: 
- Win10 pass
- Linux pass    
<details>
  <summary>Video</summary>

[no_restart_1.webm](https://user-images.githubusercontent.com/77655812/209755526-435bf4c5-6292-4462-84a6-b5743c0a5a30.webm)


</details>

Test: monerod args `--zmq-pub tcp://127.0.0.1:18083` `--disable-dns-checkpoints`    
Expected: monerod will not restart / mining begins    
Result: 
- Win10 pass
- Linux pass    
<details>
  <summary>Video</summary>

[no_restart_2.webm](https://user-images.githubusercontent.com/77655812/209755539-4c1d8371-e678-47b6-a04a-a6b907f376eb.webm)



</details>

Test: monerod args `--zmq-pub tcp://127.0.0.1:18083` `--disable-dns-checkpoints` `--no-zmq`  
Expected: monerod will restart with `--no-zmq` removed    
Result: 
- Win10 pass
- Linux pass    
<details>
  <summary>Video</summary>

[nozmq_3.webm](https://user-images.githubusercontent.com/77655812/209755574-e7062fd4-a649-4865-99dc-fef41fb8f9e3.webm)

</details>


Test: monerod args `--log-level 0` `--zmq-pub=tcp://127.0.0.1:18083`     
Expected: monerod will restart because `--disable-dns-checkpoints` not present. --log-level 0 passed also (--prune-blockchain is removed on purpose now as not neccessary) 
Result: 
- Win10 pass
- Linux pass
<details>
  <summary>Video</summary>

[log_level_4.webm](https://user-images.githubusercontent.com/77655812/209755613-3e7899dd-ff8f-4669-8f33-7ce2cebdf6ed.webm)

</details>

Test: no args    
Expected: monerod restarts with --zmq-pub and --disable-dns-checkpoints set    
Result: 
- Win10 pass   
- Linux: pass
<details>
  <summary>Video</summary>

[no_args.webm](https://user-images.githubusercontent.com/77655812/209756519-f37a8845-9512-4ae2-8d87-ffe408d6ee80.webm)


</details>

Test: monerod with only arg `--prune-blockchain`       
Expected: to restart with --prune-blockchain removed (as its not needed) but with p2pool flags set    
Result: 
- Win10 pass
- Linux pass  
<details>
  <summary>Video</summary>


[prune_removed_5.webm](https://user-images.githubusercontent.com/77655812/209755812-249424f0-df34-41b2-b262-7906579a784e.webm)

</details>

Test: monerod with only arg `--zmq-pub=tcp://127.0.0.1:18083`      
Expected: monerod to restart with zmq+disable-dns       
Result: 
- Win10 pass
- Linux pass
<details>
  <summary>Video</summary>

[zmq_eq_6.webm](https://user-images.githubusercontent.com/77655812/209755734-8dfa5682-d508-4fbd-95c0-18f1bffaa347.webm)


</details>

Test: monerod with only arg `--zmq-pub tcp://127.0.0.1:18083`      
Expected: monerod to restart with zmq+disable-dns       
Result: 
- Win10 pass
- Linux pass
<details>
  <summary>Video</summary>


[zmq_7.webm](https://user-images.githubusercontent.com/77655812/209755928-571a29a4-3d4e-4a6f-b01f-6e076287ae96.webm)


</details>

**Comment by devhyper** [comment]
Thanks for the help, I'll test again on Windows and Linux later today.

**Comment by devhyper** [comment]
Just tested all the cases on Windows (with custom data dir), all passed. Testing Linux next.

**Comment by plowsof** [comment]
I've also made another pull request to your branch with some extra tweaks that came up during review/testing. I am pleased with this PR so far!

**Comment by devhyper** [comment]
Linux tested, all cases passed. Testing omitting --disable-dns-checkpoints worked as well.

**Comment by devhyper** [comment]
@selsta Is --disable-dns-checkpoints needed for p2pool anymore?

**Comment by selsta** [comment]
@devhyper It shouldn't be necessary anymore.

**Comment by plowsof** [comment]
After squashing, i approve, all tests passing*
*win-service users will always get their daemon restarted regardless but it works fine.

Test: monerod args `--zmq-pub=tcp://127.0.0.1:18083`
Expected: monerod will not restart / mining begins    
Result: 
- Win11 pass
- Linux pass

Test: monerod args `--zmq-pub tcp://127.0.0.1:18083`   
Expected: monerod will not restart / mining begins    
Result: 
- Win11 pass
- Linux pass

Test: monerod args `--zmq-pub tcp://127.0.0.1:18083` `--no-zmq`  
Expected: monerod will restart with `--no-zmq` removed    
Result: 
- Win11 pass 
- Linux pass

Test: monerod args `--log-level 0` `--zmq-pub=tcp://127.0.0.1:18083` `--no-zmq`  
Expected: monerod will restart with --no-zmq removed and custom arg `--log-level 0` added
Result: 
- Win10 pass 
- Linux pass

Test: no args (so with defaults) 
Expected: monerod restarts with --zmq-pub set
Result: 
- Win11 pass
- Linux pass

Test: monerod with only arg `--prune-blockchain`       
Expected: to restart with --prune-blockchain removed (as its not needed) but with p2pool flags set    
Result: 
- Win11 pass
- Linux pass

Test: monerod running as a service but needs to restart
Expected: display an error with instructions
- Win11 pass (the service monerod is closed, and monero-gui spawns its own) (confirming the service is uninstalled by exit, which selsta has a PR to -monero fixing)
- Linux pass

Test: monerod running as a service with --zmq-pub in the unit file
Expected: no error / no restart 
Result:
- Win11 fail (we can only read "" args from the daemon so always must restart. if args == "" and os == win we could show a warning but not worth the hassle as win server is stopped / restarted with correct args fine, whereas systemd on linux will keep restarting it)
- Linux pass

**Comment by selsta** [comment]
@devhyper please squash

**Comment by devhyper** [comment]
@selsta @plowsof Done.

**Review by selsta — APPROVED** [review]

**Review by plowsof — APPROVED** [review]
